🎉 New Source: Exact Online

joelluijmes opened the pull request:
## What
- Implement source using Python CDK for [Exact Online](https://www.exact.com/us)
    - FYI: Exact Globe also exists, it is the same thing as Exact Online but then self-hosted.
- Closes #1596

## Recommended reading order
1. `source_exact/source.py`
2. `source_exact/streams.py`

### Bonus
`source_exact/scrape_schemas.py` - This script is not actually part of the source connector. Instead, this script automatically generates the json schemas and class source definitions. This is done by scraping the documentation pages using `requests` + `beautifulsoup`.

Not sure if it is relevant to contribute (or if it is at the right location). I thought about writing it under the `discover` function, but in this way the discovery is deterministic / isolated (and quicker).


### Reviewable notes and discussion
- This PR is rightly timed and uses the feature #19428. In that PR included is a `SingleUseRefreshTokenOauth2Authenticator` which is exactly what was needed for Exact. (🗑️  6355b517c6b8591ebedcc4700ff67c857250360c...)
    - However, as of now, the platform doesn't handle these `Control` messages. I digged up the following PR which tracks that #19811.
    - So, that PR is an impediment for this new source. I'm not sure if that prevents merging, but until then the source configuration has to be modified before each sync 😅 
- Exact comes with quite extensive list of endpoints (411 at this point). 
    - These endpoints are relatively well documented, they come with descriptions and type definitions. I used this documentation pages to automatically generate the Streams (JSON Schemas  + Class definitions). This script is also included in this PR, but not sure if relevant to add.
    - I included the property descriptions (as they help understand the data model), but opted out of the endpoint description as their descriptions are multi paragraphs of text. 
    - For now I generated the Streams for all Sync endpoints (as they are designed to be used for syncing data between third parties and Exact). In addition, I needed two other Streams (CRMClassifications and CRMClassificationNames) which are also included.

## 🚨 User Impact 🚨
No breaking change.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)


## Tests

<details><summary><strong>Unit</strong></summary>

```
Test session starts (platform: darwin, Python 3.9.11, pytest 6.2.5, pytest-sugar 0.9.6)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/.hypothesis/examples')
rootdir: /Users/joell/dev/airbyte-upstream, configfile: pytest.ini
plugins: hypothesis-6.54.6, requests-mock-1.9.3, mock-3.6.1, sugar-0.9.6, timeout-1.4.2, cov-3.0.0
collecting ... 
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_ok ✓                                                                                       3% ▍         
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_failure_forbidden ✓                                                                        5% ▌         
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_failure_missing_credentials[None] ✓                                                        8% ▊         
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_failure_missing_credentials[config1] ✓                                                    10% █         
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_failure_missing_credentials[config2] ✓                                                    13% █▍        
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_failure_missing_credentials[config3] ✓                                                    15% █▋        
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_check_connection_failure_missing_credentials[config4] ✓                                                    18% █▊        
 airbyte-integrations/connectors/source-exact/unit_tests/test_source.py::test_streams ✓                                                                                                  21% ██▏       
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_path__failure ✓                                                                                           23% ██▍       
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_path__next_page ✓                                                                                         26% ██▋       
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_path__initial ✓                                                                                           28% ██▊       
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_headers ✓                                                                                         31% ███▏      
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_next_page ✓                                                                           33% ███▍      
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_sync_stream[token-None-expected0] ✓                                                   36% ███▋      
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_sync_stream[token-123-expected1] ✓                                                    38% ███▉      
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_sync_stream[None-None-expected2] ✓                                                    41% ████▏     
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_sync_stream[None-123-expected3] ✓                                                     44% ████▍     
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_other_stream[token-None-expected0] ✓                                                  46% ████▋     
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_other_stream[token-2022-12-12T00:00:00+00:00-expected1] ✓                             49% ████▉     
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_other_stream[None-None-expected2] ✓                                                   51% █████▎    
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_other_stream[None-2022-12-12T00:00:00+00:00-expected3] ✓                              54% █████▍    
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_other_stream[None-2022-12-12T00:00:00Z-expected4] ✓                                   56% █████▋    
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_request_params_with_other_stream[None-2022-12-12T00:00:00+01:00-expected5] ✓                              59% █████▉    
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_next_page_token[body0-None] ✓                                                                             62% ██████▎   
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_next_page_token[body1-None] ✓                                                                             64% ██████▌   
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_next_page_token[body2-expected2] ✓                                                                        67% ██████▋   
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_parse_response[body0-expected0] ✓                                                                         69% ██████▉   
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_parse_response[body1-expected1] ✓                                                                         72% ███████▎  
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_parse_response[body2-expected2] ✓                                                                         74% ███████▌  
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_is_token_expired__not_expired ✓                                                                           77% ███████▊  
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_is_token_expired__expired ✓                                                                               79% ███████▉  
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_is_token_expired__failure ✓                                                                               82% ████████▎ 
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_parse_item__nested_timestamps ✓                                                                           85% ████████▌ 
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_parse_item__type_casting ✓                                                                                87% ████████▊ 
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_send_request__reraise_if_on_non_request_exception ✓                                                       90% █████████ 
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_send_request__reraise_if_response_missing_on_exception ✓                                                  92% █████████▎
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_send_request__reraise_if_not_retryable_exception ✓                                                        95% █████████▌
 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_send_request__retries_on_server_error ✓                                                                   97% █████████▊{"type": "LOG", "log": {"level": "INFO", "message": "Access token expired: will retry after refresh"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Access token expired: will retry after refresh"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Access token expired: will retry after refresh"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Access token expired: will retry after refresh"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Access token expired: will retry after refresh"}}

 airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py::test_send_request__retries_on_expired_token ✓                                                                 100% ██████████
========================================================================================== warnings summary ===========================================================================================
airbyte-integrations/connectors/source-exact/unit_tests/test_source.py: 39 warnings
airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py: 31 warnings
  /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py:43: DeprecationWarning: Call to deprecated class NoAuth. (Set `authenticator=None` instead) -- Deprecated since version 0.1.20.
    self._authenticator: HttpAuthenticator = NoAuth()

airbyte-integrations/connectors/source-exact/unit_tests/test_source.py: 39 warnings
airbyte-integrations/connectors/source-exact/unit_tests/test_streams.py: 31 warnings
  /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/.venv/lib/python3.9/site-packages/deprecated/classic.py:173: DeprecationWarning: Call to deprecated class HttpAuthenticator. (Use requests.auth.AuthBase instead) -- Deprecated since version 0.1.20.
    return old_new1(cls, *args, **kwargs)

-- Docs: https://docs.pytest.org/en/stable/warnings.html

Results (0.60s):
      39 passed
```

</details>

<details><summary><strong>Integration</strong></summary>

Not applicable.

</details>

<details><summary><strong>Acceptance</strong></summary>

```
Test session starts (platform: darwin, Python 3.9.11, pytest 6.2.5, pytest-sugar 0.9.6)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/.hypothesis/examples')
rootdir: /Users/joell/dev/airbyte-upstream, configfile: pytest.ini
plugins: hypothesis-6.54.6, requests-mock-1.9.3, mock-3.6.1, sugar-0.9.6, timeout-1.4.2, cov-3.0.0
collecting ... {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_config_match_spec[inputs0] ✓                                                                                                                   3% ▍         {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓                                                                                                                      6% ▋         {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_docker_env[inputs0] ✓                                                                                                                         10% █         {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_enum_usage[inputs0] ✓                                                                                                                         13% █▍        {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oneof_usage[inputs0] ✓                                                                                                                        16% █▋        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓                                                                                                                           19% █▉        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓                                                                                                                           23% ██▍       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓                                                                                                                         26% ██▋       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓                                                                                                         29% ██▉       {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_is_properly_marked[inputs0] ✓                                                                                                          32% ███▎      {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_defined_refs_exist_in_json_spec_file[inputs0] ✓                                                                                               35% ███▋      {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓                                                                                                              39% ███▉      {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_backward_compatibility[inputs0] s                                                                                                             42% ████▎     {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_additional_properties_is_true[inputs0] ✓                                                                                                      45% ████▌     {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓                                                                                                                        48% ████▉     {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/integration_tests/invalid_config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓                                                                                                                        52% █████▎    {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                                                                                                                      55% █████▌    {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓                                                                                               58% █████▊    {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_refs_exist_in_schema[inputs0] ✓                                                                                                  61% ██████▎   {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-allOf] ✓                                                                                         65% ██████▌   {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-not] ✓                                                                                           68% ██████▊   {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_primary_keys_exist_in_schema[inputs0] ✓                                                                                                  71% ███████▏  {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_streams_has_sync_modes[inputs0] ✓                                                                                                        74% ███████▌  {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_additional_properties_is_true[inputs0] ✓                                                                                                 77% ███████▊  {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_backward_compatibility[inputs0] s                                                                                                        81% ████████▏ {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}
{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ✓                                                                                                                          84% ████████▍ {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Docker container failed, code 1, error:\n{\"type\": \"TRACE\", \"trace\": {\"type\": \"ERROR\", \"emitted_at\": 1671032515021.95, \"error\": {\"message\": \"Something went wrong in the connector. See the logs for more details.\", \"internal_message\": \"2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\", \"stack_trace\": \"Traceback (most recent call last):\\n  File \\\"/airbyte/integration_code/main.py\\\", line 13, in <module>\\n    launch(source, sys.argv[1:])\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 131, in launch\\n    for message in source_entrypoint.run(parsed_args):\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 119, in run\\n    config_catalog = self.source.read_catalog(parsed_args.catalog)\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/source.py\\\", line 89, in read_catalog\\n    return ConfiguredAirbyteCatalog.parse_obj(self._read_json_file(catalog_path))\\n  File \\\"/usr/local/lib/python3.9/site-packages/pydantic/main.py\\\", line 521, in parse_obj\\n    return cls(**obj)\\n  File \\\"/usr/local/lib/python3.9/site-packages/pydantic/main.py\\\", line 341, in __init__\\n    raise validation_error\\npydantic.error_wrappers.ValidationError: 2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\\n\", \"failure_type\": \"system_error\"}}}\n"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_airbyte_trace_message_on_failure[inputs0] ✓                                                                                              87% ████████▊ {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}
{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓                                                                                                    90% █████████ {"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}
{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs0] ✓                                                                                                 94% █████████▍{"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}
{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_read_sequential_slices[inputs0] ✓                                                                                               97% █████████▋{"type": "LOG", "log": {"level": "INFO", "message": "Using /Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact/secrets/config.json as configuration. It is the most recent version."}}
{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs0] ✓                                                                                  100% ██████████
{"type": "LOG", "log": {"level": "INFO", "message": "/Users/joell/dev/airbyte-upstream/airbyte-integrations/connectors/source-exact - SAT run - 62be17d370bccec4ae87185e8a186590e9e5c8ae - PASSED"}}

======================================================================================================================= short test summary info ========================================================================================================================
SKIPPED [1] ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:94: The previous and actual specifications are identical.
SKIPPED [1] ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:377: The previous and actual discovered catalogs are identical.

Results (82.88s):
      29 passed
       2 skipped
```

</details>

joelluijmes commented on airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/abstract_oauth.py:
```diff
@@ -77,7 +77,7 @@ def refresh_access_token(self) -> Tuple[str, int]:
         """
         try:
             response_json = self._get_refresh_access_token_response()
-            return response_json[self.get_access_token_name()], response_json[self.get_expires_in_name()]
```
This change (and others in the CDK) are Included in #20301

sherifnada replied:
@joelluijmes can you make the base branch for this the one in #20301 to accurately understand the diff this is introducing? (it will be automatically changed to master once #20301 is merged)

joelluijmes commented on airbyte-integrations/connectors/source-exact/setup.py:
```diff
@@ -0,0 +1,36 @@
+    "source-acceptance-test",
+]
+
+DISCOVER_REQUIREMENTS = [
+    "beautifulsoup4~=4.11.1",
+    "html5lib~=1.1",
+    "pandas~=1.5.2",
+]
```
Required for the `discover_schema.py` script to automatically write all json schemas

marcosmarxm commented:
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

wissevrowl commented:
Hi @joelluijmes It's great to see a connector for Exact Online is in a pull request! We have our own Exact Online connector which we use internally. I've ran into the same issue you had with rotating access tokens. Which is why I never really created a PR. I solved this currently by saving the rotating refresh token on Blob storage in-between runs. Two points:

I see you are using the sync/deleted endpoint as a separate stream. What we did is use the Deleted endpoint to actually delete the rows from the corresponding stream. 

Another point is we noted not all data is included in Sync streams. What we did to add more data streams in the form of incremental streams is we sorted these non-sync streams by using their "Modified" date as a cursor field timestamp. This works pretty well except the deleted rows of course don't get updated until you perform a full refresh.

I am really excited to eventually switch over to an officially supported Airbyte Exact Online connector. If you want to see/copy some stuff from the code for our implementation of the Exact Online connector let me know.

joelluijmes commented:
Hi @wissevrowl, cool to hear! 

> I solved this currently by saving the rotating refresh token on Blob storage in-between runs.

Ah yes I also thought about using our own infrastructure. But that would prohibit contributing it to open-source. Luckily efforts were made to support config changes during the sync 🤓 

> I see you are using the sync/deleted endpoint as a separate stream. What we did is use the Deleted endpoint to actually delete the rows from the corresponding stream.

Interesting. Would be nice to connect to have small chat on your take then. You can mail me (me@joell.app) or find me in Airbyte slack if you are there as well.

When the connector is merged, anyone can always contribute obviously 🚀

sherifnada commented:
FYI once #3990 is done (currently in progress) the problem you mentioned with refreshing will be solved

joelluijmes commented:
@sherifnada we have been running this connector for a few weeks now. Since #19811, it works fine when refreshing the refresh token 👍 Was wondering whether I could get a review on this PR, and hopefully see it merged in near future 😇

sherifnada commented:
@marcosmarxm should be the point person for review (or should redirect to the appropriate person)!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=20480) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=20480) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: joelluijmes<br/>:x: nicklubbers<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=20480) it.</sub>

fdegier commented:
Would be great to see this merged! Happy to help.

marcosmarxm commented:
This is in the team backlog @fdegier

gvisniuc commented:
This is such a useful connector (I've been using the fork), would be awesome to have it merged!

tobiascadee commented:
+1 for other comments, great connector and can't wait for it to be available on Airbyte cloud!

joelluijmes commented:
Yes will do. I made a few attempts to keep it up to date in the past, but I eventually stopped since the pull request didn't receive any attention or review.

archangelic commented:
Completely understand! I am working on getting this reviewed now!

marcosmarxm commented:
@joelluijmes this connector is adding all services from Exact Online? (accoutning, finances, etc?) Asking because we're trying to have an integration account but each service are charge separated.

joelluijmes commented:
@marcosmarxm yes, it is integrating with the API of Exact. The API covers all the different services. I.e., doesn't really matter what kind of subscription you have, we can make integration tests with a smaller set of endpoints :)

marcosmarxm commented:
> @marcosmarxm yes, it is integrating with the API of Exact. The API covers all the different services. I.e., doesn't really matter what kind of subscription you have, we can make integration tests with a smaller set of endpoints :)

Asking because each service is charged individually. For most cases you use all of them? Not sure if we should split each service into a separate connector. I'll ask the connector team and return to you.

gvisniuc commented:
> > @marcosmarxm yes, it is integrating with the API of Exact. The API covers all the different services. I.e., doesn't really matter what kind of subscription you have, we can make integration tests with a smaller set of endpoints :)
> 
> Asking because each service is charged individually. For most cases you use all of them? Not sure if we should split each service into a separate connector. Saying this maybe someone will try only the Finance and the discover schema fail because doesn't have permission to the others one. I'll ask the connector team and return to you.

If I can also comment on this, since I use this connect in its current state. From personal experience you use the full ExactOnline stack since you want to link the various entities (transactions, suppliers, with GL accounts etc).

joelluijmes commented:
> Asking because each service is charged individually. For most cases you use all of them? Not sure if we should split each service into a separate connector. Saying this maybe someone will try only the Finance and the discover schema fail because doesn't have permission to the others one. I'll ask the connector team and return to you.

Agreeing with @gvisniuc. I don't think it makes sense (or is useful) to have separate connectors for each individual service. In general you have the bulk of the services. 

However, I understand your concern about ensuring a good user experience and avoiding strange errors when a particular service is not available. For instance, in the current implementation, it's possible that not all Exact users have the Subscription service enabled, yet its streams are included in this pull request.

Looking bit back in the code, I noticed that the configuration is passed to the discover function. One approach we could take is to introduce a list of enabled Exact services and associate it with the included streams. This way, we can address the issue.

marcosmarxm commented on airbyte-integrations/connectors/source-exact/source_exact/streams.py:
```diff
@@ -0,0 +1,465 @@
+        division = str(stream_slice["division"])
+        self._url_base = f"https://start.exactonline.nl/api/v1/{division}/"
+
+        self.logger.info(f"Syncing division {division}...")
+
+        # Reset state for full refresh
+        if sync_mode == SyncMode.full_refresh:
+            self._state_per_division[division] = {}
+
+        for record in super().read_records(sync_mode=sync_mode, stream_slice=stream_slice, **kwargs):
+            # Track the largest cursor value
+            if self.cursor_field and sync_mode == SyncMode.incremental:
+                cursor_value = record[self.cursor_field]
+                current_cursor_value = self._state_per_division[division].get(self.cursor_field)
+                current_cursor_value = cursor_value if not current_cursor_value else current_cursor_value
+
+                if current_cursor_value:
+                    self._state_per_division[division].update({self.cursor_field: max(cursor_value, current_cursor_value)})
+
+            yield record
```
Can you explain why you're overriding the base function?

joelluijmes replied:
It was documented in the pydoc, but I refactored it a bit I hope it is clearer now why its needed.

marcosmarxm commented:
Close due inactivity.

joelluijmes commented:
Oh sorry, I think I didn't see (or forgot about) a review was posted. Regardless, I'm definitely determined to get this merged. 😬
Right now I'm on holiday so I won't get to it for few weeks. But when I have the time I'll reopen the PR or submit a new PR.

gvisniuc commented:
@joelluijmes any chance this gets picked up again in the future? Would be awesome 💪

joelluijmes commented on airbyte-integrations/connectors/source-exact/source_exact/source.py:
```diff
@@ -0,0 +1,128 @@
+        if not divisions:
+            return False, "Missing divisions"
+
+        # TODO: check with airbyte whether control messages are handled during connection check (for token refresh)
+        # try:
+        #     headers = {
+        #         "Authorization": f"Bearer {access_token}",
+        #         "Accept": "application/json",
+        #     }
+
+        #     response = requests.get(
+        #         "https://start.exactonline.nl/api/v1/current/Me",
+        #         headers=headers,
+        #         timeout=15,
+        #     )
+
+        #     response.raise_for_status()
+        #     logger.info(f"Connection check successful. Details:\n{json.dumps(response.json())}")
+        # except requests.RequestException as exc:
+        #     return (
+        #         False,
+        #         f"Exception happened during connection check. Validate that the access_token is still valid at this point. Details\n{exc}",
+        #     )
+
```
There is a TODO note which should be resolved. Exact only allows single use refresh tokens, so it could be possible that this health check triggers a new token. If that is the case, will the token be persisted then in the config? 

> Is the /current/Me return any information about what services the account has access to it?
Nope, just some metadata about the user like the name etc.

joelluijmes commented on airbyte-integrations/connectors/source-exact/source_exact/source.py:
```diff
@@ -0,0 +1,128 @@
+
+        return True, None
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        return [
+            SyncCashflowPaymentTerms(config),
+            SyncCRMAccounts(config),
```
Alright, there is no endpoint or something like scopes to do this. However, I can make a "test" request during discover. When the user does not have access to a stream (Exact feature), it returns a 403 Forbidden.

joelluijmes commented on airbyte-integrations/connectors/source-exact/source_exact/streams.py:
```diff
@@ -0,0 +1,465 @@
+        def parse_value(key, value):
+            if isinstance(value, dict):
+                return {k: parse_value(k, v) for k, v in value.items()}
+
+            if isinstance(value, list):
+                return [parse_value(key, v) for v in value]
+
+            if isinstance(value, str):
+                # Exact returns timestamps in following format: /Date(1672531200000)/ (OData date format).
+                # The value is in seconds since Epoch (UNIX time) format. However NOTE that the time is in Dutch timezone:
+                # - CET (UTC+01.00) in the winter
+                # - CET (UTC+02.00) in the summer
+                # The `pendulum.from_timestamp(x, 'CET')` takes into account both summer and winter time.
+                # A timestamp x in the winter is read as UTC+01.00 and a timestamp x in the summer is read as UTC+02.00.
+                #
+                # Exact API docs: https://support.exactonline.com/community/s/knowledge-base#All-All-DNO-Content-faq-rest-api
+                match_timestamp = regex_timestamp.match(value)
+                if match_timestamp:
+                    cet_unix_seconds = int(match_timestamp.group(1)) / 1000
+                    cet_offset = pendulum.from_timestamp(cet_unix_seconds, 'CET').offset # either 3600 or 7200
+
+                    # Convert CET (+1 or +2) to UTC
+                    utc_unix_seconds = cet_unix_seconds - cet_offset
+                    
+                    # Create timestamp in UTC
+                    timestamp = pendulum.from_timestamp(utc_unix_seconds)
+
+                    return timestamp.isoformat()
+
+                expected_type = property_type_lookup.get(key)
+                if expected_type == "number":
+                    return float(value)
+                elif expected_type == "integer":
+                    return int(value)
+                elif expected_type == "boolean":
+                    return value and value.lower() == "true"
+
+            return value
+
+        return parse_value(None, obj)
```
Its actually already in tested: 
https://github.com/airbytehq/airbyte/pull/20480/files#diff-51d27bf2f8f624347fd91d702e28e6fd1fd7996c578f995fcb9cce99376a9b67R166-R206

joelluijmes commented:
Alright, I finally found some time to finish this PR 🥳 

@marcosmarxm can you reopen the PR, or should I create a new PR? I believe I addressed all your comments, and/or replied on Github.

Jeroendevr commented:
@marcosmarxm Here a heads up for another reopening.

natikgadzhi commented:
Hey! @Jeroendevr if you're down to make this connector work, I'm happy to help.

But, since 2022 and 2023, we've tweaked the CDK quite a lot, so it could be easier to start from scratch.

I would also love to see if we could make this work in low-code instead of Python. Is the list of streams itself fixed? Are the stream schemas dynamic?

Jeroendevr commented:
Hey @natikgadzhi, 

_LowCode CDK_
The API itself is quite extensive so I don't exactly know if the list of streams is fixed. @joelluijmes Do you know it perhaps?
I have worked with the lowcode CDK for some sources and it works generally well but I do not know if it can be used for this source. 

_Updated CDK_
Do you suggest that a (sort of) rebase on the current HEAD would require more work than rewriting the the changes from this PR into the current repo? 

As a first step I am leaning more towards implementing this PR (or a rewrite/rebase of it) than a LowCode CDK as I know this PR works but I cannot forsee if the same implementation is possible in LowCode CDK. Would that be ok? Perhaps later if I or @joelluijmes has enough confidence in LowCode CDK we can implement it though that platform.

natikgadzhi commented:
Looking at 67 files, it's very likely that it would be much easier to start with a low-code connector on CDK 4.5 from scratch and copy and paste pieces from here.

natikgadzhi commented:
> I've had some communication with @marcosmarxm over Slack. The latest update is that I plan to schedule a call with Teo to explore the possibility of a partnership between Gynzy and Airbyte. This partnership would involve us taking responsibility for maintaining this custom connector.

That sounds great! Especially if the connector you have works well for you, and you're willing to answer questions about it, and dust it off from time to time and keep it on the most recent minor CDK version, sure.
